### DISCUSSION TITLE
brew test-bot action fails at brew test-bot --only-formulae step

### DISCUSSION THREAD

**Discussion by cesarcoatl** [Homebrew/discussions#3089] [Tap maintenance and Homebrew development]
I maintain a [tap](https://github.com/coatl-dev/homebrew-coatl-dev), and a few months back I noticed I had to add some more logic to the `post_install` block as the `brew test-bot --only-formulae` was exiting with code 1.

Before:
```ruby
  def post_install
    %w[Notice.txt README.txt].each do |f|
      libexec.install "#{prefix}/#{f}"
    end
```

Now:
```ruby
  def post_install
    %w[Notice.txt README.txt].each do |f|
      libexec.install "#{prefix}/#{f}" if File.exist?("#{prefix}/#{f}")
    end
```

I have verified that if I don't move those files, they remain at the root directory for the formula, but I need them at the `libexec` directory.

The formula installs fine, but it is the step at the `tests.yml` workflow the one that fails.

Failed workflow [here](https://github.com/coatl-dev/homebrew-coatl-dev/runs/5575187347?check_suite_focus=true#step:8:47)

Successful workflow after adding the `File.exist?("#{prefix}/#{f}")` condition, [here](https://github.com/coatl-dev/homebrew-coatl-dev/runs/5575455071?check_suite_focus=true#step:8:47).

For that same formula, https://github.com/coatl-dev/homebrew-coatl-dev/pull/26 didn't fail back in June 2021, but I did have to put that same condition for another formula back in December 2021 (https://github.com/coatl-dev/homebrew-coatl-dev/pull/35).

So, what changed for `brew test-bot` between June and December?

Thanks!

**Comment by gromgit** [comment]
When I unpack the zip, I see this:
```
$ ls
com.inductiveautomation.ignition.plist  data/  gcu.sh  gwcmd.sh  ignition-gateway  ignition.sh  lib/  logs/  Notice.txt  README.txt  temp/  user-lib/  webserver/  wrapper-license.conf
```
and your `install` block begins with:
```
  def install
    # Relocate data
    mv "data", "ignition7.9"
    etc.install "ignition7.9" unless (etc/"ignition7.9").exist?
    rm_rf "ignition7.9"


    # Relocate logs
    mv "logs", "ignition7.9"
    var.install "ignition7.9" unless (var/"ignition7.9").exist?
    rm_rf "ignition7.9"


    # Install
    libexec.install Dir["*"]
```
Unless I'm misreading it, your formula has already moved the two `*.txt` files into `libexec`, so I'm not surprised your original `post_install` block errored out.

You should confirm it by building a bottle manually and checking its contents, but I'm pretty sure that's where things went south.

**Reply by cesarcoatl** [reply]
Unfortunately `libexec.install Dir["*"]` does not "install" those files I'm moving at `post_install`.

If you check other formulae, some files are "left behind" at the root directory for the formula.

**Comment by carlocab** [comment]
> So, what changed for `brew test-bot` between June and December?

Nothing related, probably. Still, you can [see for yourself](https://github.com/Homebrew/homebrew-test-bot/compare/master@{10.months.ago}..master).

**Comment by cesarcoatl** [comment]
I modified the `post_install` block on my local tap:

```ruby
  def post_install
    # Unzip the new runtime
    system bin/"ignition", "checkruntimes"

    # Update ignition.conf
    system bin/"ignition", "runupgrader"
  end
```

Uninstalled it on my machine and installed again, and as you can see, the files I was moving at `post_install` are found at the root directory for the formula:

```bash
$ export HOMEBREW_NO_AUTO_UPDATE=1
$ brew install coatl-dev/coatl-dev/ignition
==> Downloading https://files.inductiveautomation.com/release/ia/8.1.15/20220301-1426/Ignition-macOs-x86-64-8.1.15.zip
Already downloaded: /Users/thecesrom/Library/Caches/Homebrew/downloads/5a5266a219ab98640023b300c47ffe0a2ad8b454d83edfac3b657b39e8a95d3f--Ignition-macOs-x86-64-8.1.15.zip
==> Installing ignition from coatl-dev/coatl-dev
==> /usr/local/Cellar/ignition/8.1.15/bin/ignition checkruntimes
==> /usr/local/Cellar/ignition/8.1.15/bin/ignition runupgrader
==> Caveats
The data and logs folders have been symlinked to:
  data: /usr/local/etc/ignition
  logs: /usr/local/var/ignition

To start coatl-dev/coatl-dev/ignition now and restart at login:
  brew services start coatl-dev/coatl-dev/ignition
==> Summary
🍺  /usr/local/Cellar/ignition/8.1.15: 1,409 files, 1.5GB, built in 53 seconds
==> Running `brew cleanup ignition`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
$ cd "$(brew --prefix ignition)"
$ ls
INSTALL_RECEIPT.json         Notice.txt                   bin                          libexec
License.html                 README.txt                   homebrew.mxcl.ignition.plist
```

Evidently `libexec.install Dir["*"]` does not install everything, for some reason.

I restored the formula to the original `post_install`, and reinstalled:

```ruby
  def post_install
    # Relocate files
    %w[License.html Notice.txt README.txt].each do |f|
      libexec.install "#{prefix}/#{f}" if File.exist?("#{prefix}/#{f}")
    end

    # Unzip the new runtime
    system bin/"ignition", "checkruntimes"

    # Update ignition.conf
    system bin/"ignition", "runupgrader"
  end
```

```bash
$ brew install coatl-dev/coatl-dev/ignition
...
$ cd "$(brew --prefix ignition)"
$ ls
INSTALL_RECEIPT.json         bin                          homebrew.mxcl.ignition.plist libexec
```

Only that way those files are moved.

So providing those two workflows, one failing if the `if File.exist?("#{prefix}/#{f}")` condition is not present, and the one succeeding when it's there. I'm still wondering what changed.

And why does `libexec.install Dir["*"]` not install everything on `libexec`?

Thanks!

**Reply by gromgit** [reply]
OK, something's gone weird. I applied the following patch to your formula:
```
diff --git a/Formula/ignition.rb b/Formula/ignition.rb
index e736653..0563da7 100644
--- a/Formula/ignition.rb
+++ b/Formula/ignition.rb
@@ -25,6 +25,7 @@ class Ignition < Formula
     rm_rf "ignition"
 
     # Install
+    ohai "Installing #{Dir["*"]} in libexec"
     libexec.install Dir["*"]
 
     # Make files executable
```
and tried to build a bottle:
```
$ brew install --verbose --debug --build-bottle coatl-dev/coatl-dev/ignition
[...]
==> Installing ["temp", "gwcmd.sh", "License.html", "user-lib", "webserver", "Notice.txt", "ignition-gateway", "ignition-util.sh", "lib", "README.txt", "wrapper-license.conf", "com.inductiveautomation.ignition.plist", "ignition.sh"] in libexec
/usr/local/Homebrew/Library/Homebrew/shims/shared/git --version
==> Cleaning
==> Finishing up
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/coatl-dev/homebrew-coatl-dev/Formula/ignition@7.9.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/coatl-dev/homebrew-coatl-dev/Formula/ignition@8.0.rb
ln -s ../Cellar/ignition/8.1.15/bin/ignition ignition
==> Changing dylib ID of /usr/local/Cellar/ignition/8.1.15/libexec/lib/core/gateway/libignition.dylib
  from libignition.jnilib
    to /usr/local/opt/ignition/libexec/lib/core/gateway/libignition.jnilib
==> Not running 'post_install' as we're building a bottle
You can run it manually using:
  brew postinstall coatl-dev/coatl-dev/ignition
==> Caveats
The data and logs folders have been symlinked to:
  data: /usr/local/etc/ignition
  logs: /usr/local/var/ignition

To start coatl-dev/coatl-dev/ignition now and restart at login:
  brew services start coatl-dev/coatl-dev/ignition
==> Summary
🍺  /usr/local/Cellar/ignition/8.1.15: 1,319 files, 1.4GB, built in 1 minute 8 seconds
==> Running `brew cleanup ignition`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
[...]
==> Installation times
ignition                 63.887 s

$ ls /usr/local/Cellar/ignition/8.1.15/
.bottle/  .brew/  INSTALL_RECEIPT.json  License.html  Notice.txt  README.txt  bin/  homebrew.mxcl.ignition.plist@  libexec/
```
So `Dir["*"]` definitely expanded to the correct list of filenames, but `libexec.install` seems to ignore some of them, no idea why.

In any case, it's pretty clear that it's not a `brew test-bot` issue, since manual builds fail in the same way.

**Reply by cesarcoatl** [reply]
Trying to rule out any error on my formula, I installed `jenv` and I see the same behavior; not all files are installed into `libexec`.

```bash
$ brew install jenv
...
$ cd "$(brew --prefix jenv)"
$ ls
INSTALL_RECEIPT.json README.md            libexec
LICENSE              bin                  share
```

Is it a `brew` bug then?
